extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/TestsAsBeansProcessor.java

```diff
@@ -31,4 +46,92 @@ public void testClassBeans(List<TestClassBeanBuildItem> items, BuildProducer<Add
         producer.produce(builder.build());
     }
 
+    @BuildStep(onlyIf = IsTest.class)
+    AnnotationsTransformerBuildItem vetoTestProfileBeans(Optional<TestProfileBuildItem> testProfile,
+            CustomScopeAnnotationsBuildItem customScopes, CombinedIndexBuildItem index) {
+        if (index.getIndex().getAllKnownImplementors(QUARKUS_TEST_PROFILE).isEmpty()) {
+            // No test profiles found
+            return null;
+        }
+
+        Set<DotName> currentTestProfileHierarchy = initTestProfileHierarchy(testProfile, index.getComputingIndex());
+        return new AnnotationsTransformerBuildItem(new AnnotationsTransformer() {
+
+            @Override
+            public void transform(TransformationContext context) {
+                AnnotationTarget target = context.getTarget();
+                if (target.kind() == Kind.METHOD) {
+                    vetoProducerIfNecessary(target.asMethod().declaringClass(), context);
+                } else if (target.kind() == Kind.FIELD) {
+                    vetoProducerIfNecessary(target.asField().declaringClass(), context);
+                } else if (target.kind() == Kind.CLASS) {
+                    ClassInfo clazz = target.asClass();
+                    if (clazz.nestingType() == NestingType.INNER && Modifier.isStatic(clazz.flags())) {
+                        ClassInfo enclosing = index.getComputingIndex().getClassByName(clazz.enclosingClass());
+                        if (customScopes.isScopeIn(context.getAnnotations())
+                                && isTestProfileClass(enclosing, index.getComputingIndex())
```

Comment:
Technically this will only veto nested classes of test profile classes, but not nested classes of nested classes of test profile classes :-), but I guess that's not a big deal.

Reply:
Yes, nested classes of nested classes of test profile classes are not supported :shrug: